File: GLLara/GLLBoneListController.m
//
//  GLLBoneListController.m
//  GLLara
//
//  Created by Torsten Kammer on 13.01.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

#import "GLLBoneListController.h"

#import "NSArray+Map.h"
#import "GLLItem.h"
#import "GLLBoneController.h"

@interface GLLBoneListController ()

@property (nonatomic) NSArray<GLLBoneController *> *rootBoneControllers;
@property (nonatomic, readwrite) NSMutableArray<GLLBoneController *> *boneControllers;

@end

@implementation GLLBoneListController

- (id)initWithItem:(GLLItem *)item outlineView:(NSOutlineView *)outlineView parent:(id)parentController;
{
	if (!(self = [super init])) return nil;
	
	self.item = item;
	self.parentController = parentController;
	_outlineView = outlineView;
	
	self.boneControllers = [[self.item.combinedBones map:^(GLLItemBone *bone){
		return [[GLLBoneController alloc] initWithBone:bone listController:self];
	}] mutableCopy];
	self.rootBoneControllers = [self.boneControllers filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"bone.parent == NULL"]];
	
	[self.item addObserver:self forKeyPath:@"childItems" options:NSKeyValueObservingOptionNew | NSKeyValueObservingOptionOld | NSKeyValueObservingOptionInitial context:NULL];
	
	return self;
}

- (void)dealloc
{
	[self.item removeObserver:self forKeyPath:@"childItems"];
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
	if ([keyPath isEqual:@"childItems"])
	{
		// Observe the children of the new items, too, and remove observing for old ones
		if (![change[NSKeyValueChangeOldKey] isKindOfClass:[NSNull class]])
			for (GLLItem *item in change[NSKeyValueChangeOldKey])
				[item removeObserver:self forKeyPath:@"childItems"];
		
		if (![change[NSKeyValueChangeNewKey] isKindOfClass:[NSNull class]])
			for (GLLItem *item in change[NSKeyValueChangeNewKey])
				[item addObserver:self forKeyPath:@"childItems" options: NSKeyValueObservingOptionInitial context:NULL];
		
		// Run only after delay, to ensure that everything has been set correctly by the time this code gets active.
		dispatch_async(dispatch_get_main_queue(), ^{
			NSOrderedSet<GLLItemBone *> *newBones = self.item.combinedBones;
			
			// Find existing bones
			NSOrderedSet<GLLItemBone *> *existingBones = [NSOrderedSet orderedSetWithArray:[self.boneControllers valueForKeyPath:@"bone"]];
			
			// Remove bone controllers not in the old list
			NSMutableOrderedSet<GLLItemBone *> *deleted = [NSMutableOrderedSet orderedSetWithOrderedSet:existingBones];
			[deleted minusOrderedSet:newBones];
			
			NSIndexSet *indicesOfDeletedBoneControllers = [self.boneControllers indexesOfObjectsPassingTest:^(GLLBoneController *boneController, NSUInteger idx, BOOL *stop){
				return [deleted containsObject:boneController.bone];
			}];
			[self.boneControllers removeObjectsAtIndexes:indicesOfDeletedBoneControllers];
			
			// Add bone controllers in the new list
			NSMutableOrderedSet<GLLItemBone *> *added = [NSMutableOrderedSet orderedSetWithOrderedSet:newBones];
			[added minusOrderedSet:existingBones];
			[self.boneControllers addObjectsFromArray:[added map:^(GLLItemBone *bone){
				return [[GLLBoneController alloc] initWithBone:bone listController:self];
			}]];
			
			// Update
			self.rootBoneControllers = [self.boneControllers filteredArrayUsingPredicate:[NSPredicate predicateWithFormat:@"bone.parent == NULL"]];
			[self.outlineView reloadItem:self reloadChildren:YES];
		});
	}
}

- (NSArray *)allSelectableControllers
{
	return self.boneControllers;
}

#pragma mark - Outline View Data Source

- (id)outlineView:(NSOutlineView *)outlineView child:(NSInteger)index ofItem:(id)item
{
	return self.rootBoneControllers[index];
}

- (id)outlineView:(NSOutlineView *)outlineView objectValueForTableColumn:(NSTableColumn *)tableColumn byItem:(id)item
{
	return NSLocalizedString(@"Bones", @"source view header");
}

- (BOOL)outlineView:(NSOutlineView *)outlineView isItemExpandable:(id)item
{
	return YES;
}

- (NSInteger)outlineView:(NSOutlineView *)outlineView numberOfChildrenOfItem:(id)item
{
	return self.rootBoneControllers.count;
}

#pragma mark - Outline View Delegate

- (BOOL)outlineView:(NSOutlineView *)outlineView shouldSelectItem:(id)item
{
	return NO;
}

@end

File: GLLara/GLLBoneListController.swift
//
//  GLLBoneListController.m
//  GLLara
//
//  Created by Torsten Kammer on 13.01.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

import Cocoa

/*!
 * Source list controller for a list of bones. Child of an item.
 * Also includes the bones of subitems. It creates all bone controllers, not
 * just the root bones. The bone controllers then access it to get their
 * children.
 */
@objc class GLLBoneListController: NSObject, NSOutlineViewDataSource, NSOutlineViewDelegate {
    init(item: GLLItem, outlineView: NSOutlineView, parent: AnyObject) {
        self.item = item
        self.outlineView = outlineView
        self.parentController = parent
        
        super.init()
        
        boneControllers = item.combinedBones().map { GLLBoneController(bone: ($0 as! GLLItemBone), listController: self) }
        rootBoneControllers = boneControllers.filter { $0.bone.parent == nil }
        
        startObservingChildItems(of: item)
    }
    
    private func startObservingChildItems(of item: GLLItem) {
        observations[item] = item.observe(\GLLItem.childItems, options: [.new, .old, .initial]) { _, change in
            for item in (change.oldValue ?? []) ?? [] {
                self.stopObservingChildItems(of: item as! GLLItem)
            }
            for item in (change.newValue ?? []) ?? [] {
                self.startObservingChildItems(of: item as! GLLItem)
            }
            
            // Run only after delay, to ensure that everything has been set correctly by the time this code gets active.
            DispatchQueue.main.async { [self] in
                boneControllers = item.combinedBones().map { GLLBoneController(bone: ($0 as! GLLItemBone), listController: self) }
                rootBoneControllers = boneControllers.filter { $0.bone.parent == nil }
                
                outlineView.reloadItem(self, reloadChildren: true)
            }
        }
    }
    
    private func stopObservingChildItems(of item: GLLItem) {
        observations.removeValue(forKey: item)
        for child in item.childItems {
            stopObservingChildItems(of: child as! GLLItem)
        }
    }
    
    @objc var item: GLLItem
    @objc var allSelectableControllers: [GLLBoneController] {
        return boneControllers;
    }
    @objc var boneControllers: [GLLBoneController] = []
    @objc let outlineView: NSOutlineView
    @objc weak var parentController: AnyObject?
    
    @objc private var rootBoneControllers: [GLLBoneController] = []
    
    private var observations: [GLLItem : NSKeyValueObservation] = [:]
    
    func outlineView(_ outlineView: NSOutlineView, child index: Int, ofItem item: Any?) -> Any {
        return rootBoneControllers[index]
    }
    
    func outlineView(_ outlineView: NSOutlineView, objectValueFor tableColumn: NSTableColumn?, byItem item: Any?) -> Any? {
        return NSLocalizedString("Bones", comment: "source view header")
    }
    
    func outlineView(_ outlineView: NSOutlineView, isItemExpandable item: Any) -> Bool {
        return true
    }
    
    func outlineView(_ outlineView: NSOutlineView, numberOfChildrenOfItem item: Any?) -> Int {
        return rootBoneControllers.count
    }
    
    func outlineView(_ outlineView: NSOutlineView, shouldSelectItem item: Any) -> Bool {
        return false
    }
}
